File: js/scripts/cochran.js
/**
 * cochran.js
 * -----------------------------------------------------------------
 * Portage JS de cochran.py — calcul de la taille d'échantillon requise
 * (formule de Cochran, avec correction pour population finie).
 * Purement arithmétique, aucune dépendance externe : portage 1:1.
 */

(function () {
    // Mêmes valeurs par défaut que le script Python (issues de la thèse)
    const DEFAULTS = {
        N: 9958,
        Z: 1.96,
        p: 0.5,
        e: 0.05
    };

    function renderForm() {
        return `
        <div class="grid grid-cols-1 md:grid-cols-2 gap-6 mb-6">
            <div class="input-group">
                <label class="text-xs mb-1 block opacity-80">> Taille totale de la population (N)</label>
                <input type="number" id="ck-N" class="terminal-input" placeholder="${DEFAULTS.N}" value="${DEFAULTS.N}">
            </div>
            <div class="input-group">
                <label class="text-xs mb-1 block opacity-80">> Écart-réduit (Z) — 1.96 pour 95%</label>
                <input type="number" id="ck-Z" class="terminal-input" step="0.01" placeholder="${DEFAULTS.Z}" value="${DEFAULTS.Z}">
            </div>
            <div class="input-group">
                <label class="text-xs mb-1 block opacity-80">> Proportion présumée (p) — 0.5 = variance max</label>
                <input type="number" id="ck-p" class="terminal-input" step="0.01" min="0" max="1" placeholder="${DEFAULTS.p}" value="${DEFAULTS.p}">
            </div>
            <div class="input-group">
                <label class="text-xs mb-1 block opacity-80">> Marge d'erreur tolérée (e) — 0.05 = 5%</label>
                <input type="number" id="ck-e" class="terminal-input" step="0.01" placeholder="${DEFAULTS.e}" value="${DEFAULTS.e}">
            </div>
        </div>
        <button id="run-btn" class="terminal-btn w-full">EXECUTE --cochran-sample-size</button>`;
    }

    async function run(scrollback) {
        const N = parseFloat(document.getElementById("ck-N").value);
        const Z = parseFloat(document.getElementById("ck-Z").value);
        const p = parseFloat(document.getElementById("ck-p").value);
        const e = parseFloat(document.getElementById("ck-e").value);

        if ([N, Z, p, e].some(isNaN)) {
            scrollback.print("[ERROR] Veuillez entrer des valeurs numériques valides pour N, Z, p et e.", "tl-error");
            return;
        }

        if (p < 0 || p > 1) {
            scrollback.print("[ERROR] La proportion présumée (p) doit être comprise entre 0 et 1.", "tl-error");
            return;
        }

        if (e <= 0) {
            scrollback.print("[ERROR] La marge d'erreur (e) doit être strictement positive.", "tl-error");
            return;
        }

        const q = 1 - p;

        await scrollback.thinking("calcul de l'échantillon théorique (n₀)");

        // n0 = (Z^2 * p * q) / e^2
        const n0 = (Math.pow(Z, 2) * p * q) / Math.pow(e, 2);

        // n = n0 / (1 + (n0 - 1) / N)
        const nFinal = n0 / (1 + (n0 - 1) / N);

        scrollback.print("", "");
        scrollback.print("[RÉSULTATS]", "tl-highlight");
        scrollback.print(`• Variance maximale estimée (p×q)         : ${(p * q).toFixed(4)}`, "tl-info");
        scrollback.print(`• Échantillon théorique brut (n₀)         : ${n0.toFixed(4)} lignes`, "tl-info");
        scrollback.print(`• Échantillon minimal requis réajusté (n) : ${nFinal.toFixed(2)} lignes`, "tl-success");
        scrollback.print(`• Taille à retenir (arrondi supérieur)    : ${Math.ceil(nFinal)} lignes`, "tl-success");
    }

    registerTerminalScript({
        id: "cochran",
        pill: "cochran.js",
        command: "./cochran.py",
        ready: true,
        intro: [
            { text: "Portage JS de cochran.py — taille d'échantillon requise (formule de Cochran, correction population finie).", cls: "tl-comment" },
            { text: "Valeurs par défaut pré-remplies d'après la méthodologie de la thèse (N=9958, Z=1.96, p=0.5, e=0.05).", cls: "tl-comment" }
        ],
        renderForm,
        run
    });
})();

File: py_scripts/cochran.py
import math


def calculer_echantillon_cochran():
    print("=== CALCULATEUR METHODOLOGIQUE : FORMULE DE COCHRAN ===")
    print("Ajustement de l'échantillon pour une population finie\n")

    # 1. Collecte des paramètres avec valeurs par défaut de la thèse
    try:
        # Taille totale du corpus (N)
        N_input = input("-> Taille totale de la population (N) [Défaut: 9958] : ")
        N = int(N_input) if N_input.strip() else 9958

        # Score Z (Z-score lié au niveau de confiance)
        Z_input = input("-> Valeur de l'écart-réduit (Z) [Défaut: 1.96 pour 95%] : ")
        Z = float(Z_input) if Z_input.strip() else 1.96

        # Proportion présumée (p)
        p_input = input(
            "-> Proportion présumée (p) [Défaut: 0.5 pour variance max] : "
            )
        p = float(p_input) if p_input.strip() else 0.5
        q = 1 - p

        # Marge d'erreur tolérée (e)
        e_input = input("-> Marge d'erreur tolérée (e) [Défaut: 0.05 pour 5%] : ")
        e = float(e_input) if e_input.strip() else 0.05

    except ValueError:
        print("\n[Erreur] Veuillez entrer des valeurs numériques valides.")
        return

    # 2. Calcul de l'échantillon théorique pour population infinie (n0)
    # Formule : n0 = (Z^2 * p * q) / e^2
    n0 = (pow(Z, 2) * p * q) / pow(e, 2)

    # 3. Calcul de l'échantillon corrigé pour population finie (n)
    # Formule : n = n0 / (1 + (n0 - 1) / N)
    n_final = n0 / (1 + ((n0 - 1) / N))

    # 4. Affichage des résultats
    print("\n" + "-" * 50)
    print("RÉSULTATS DE L'ANALYSE STATISTIQUE :")
    print("-" * 50)
    print(f"• Variance maximale estimée (p*q) : {p * q}")
    print(f"• Échantillon théorique brut (n₀) : {n0:.4f} lignes")
    print(f"• Échantillon minimal requis réajusté (n) : {n_final:.2f} lignes")
    print(f"• Taille à retenir (arrondi supérieur)   : {math.ceil(n_final)} lignes")
    print("-" * 50)


# Exécution du script
if __name__ == "__main__":
    calculer_echantillon_cochran()